thrivecoin/sources/reward.move

```diff
@@ -0,0 +1,166 @@
+module thrivecoin::reward {
+  // imports
+  use sui::object::{Self, UID};
+  use sui::transfer;
+  use sui::tx_context::{Self, TxContext};
+  use sui::vec_set::{Self, VecSet};
+  use sui::table::{Self, Table};
+  use sui::sui::SUI;
+  use sui::coin::{Self, Coin};
+  use sui::balance::{Self, Balance};
+  use sui::types;
+
+  // errors
+  const ENotOneTimeWitness: u64 = 1;
+  const ENotWriter: u64 = 2;
+  const ETreasuryInsufficient: u64 = 3;
+  const EBalInsufficient: u64 = 4;
+
+  // structures
+  struct AdminRole has key {
+    id: UID
+  }
+
+  struct WriterRole has key {
+    id: UID,
+    list: VecSet<address>
+  }
+  
+  struct RewardLedger has key, store {
+    id: UID,
+    treasury: Balance<SUI>,
+    total_rewards: u64,
+    balances: Table<address, u64>
+  }
+
+  // OTW
+  struct REWARD has drop {}
+
+  // initializer
+  fun init(otw: REWARD, ctx: &mut TxContext) {
+    assert!(types::is_one_time_witness(&otw), ENotOneTimeWitness);
+
+    transfer::transfer(AdminRole {
+      id: object::new(ctx)
+    }, tx_context::sender(ctx));
+
+    transfer::share_object(WriterRole {
+      id: object::new(ctx),
+      list: vec_set::singleton(tx_context::sender(ctx))
+    });
+
+    transfer::share_object(RewardLedger {
+      id: object::new(ctx),
+      treasury: balance::zero(),
+      total_rewards: 0,
+      balances: table::new<address, u64>(ctx)
+    });
+  }
+
+  // role functions
+  public fun transfer_admin_role (admin_role: AdminRole, new_owner: address) {
+    transfer::transfer(admin_role, new_owner);
+  }
+
+  public fun add_writer (_: &AdminRole, writer_role: &mut WriterRole, account: address) {
+    vec_set::insert(&mut writer_role.list, account);
+  }
+
+  public fun del_writer(_: &AdminRole, writer_role: &mut WriterRole, account: address) {
+    vec_set::remove(&mut writer_role.list, &account);
+  }
+
+  public fun writer_list(self: &WriterRole): VecSet<address> { self.list }
+
+  // reward functions
+  public fun deposit (
+    reward_ledger: &mut RewardLedger,
+    coin: &mut Coin<SUI>,
+    amount: u64
+  ) {
+    let coin_balance = coin::balance_mut(coin);
+    let payment = balance::split(coin_balance, amount);
+    balance::join(&mut reward_ledger.treasury, payment);
+  }
+
+  public fun add_reward (
```

Comment:
It seems there's a potential issue where writers can give rewards exceeding the treasury balance. For example, if the treasury balance is 100 and recipients A and B have balances of 50 and 60, respectively, when recipient A claims their rewards, the treasury balance becomes 50. This could lead to inefficiencies when recipient B attempts to claim their rewards, as the treasury balance is insufficient.

Does this align with your desired requirements? If not, you might consider adding an assertion in the `add_reward` function to check whether the total rewards will not exceed the treasury balance, before adding a new reward. This ensures that rewards are given within the available treasury balance and prevents potential discrepancies in reward claims.

Reply:
In our case we usually first store rewards and after that we calculate total amount needed and top up the contract with that amount (treasury in this case). So technically is possible to end up in state where contract has insufficient funds, but we ensure that we have enough funds before adding the rewards and topping up the contract.

Additionally we have withdraw_treasury that allows withdrawing any amount that exceeds total rewards, this is a mechanism we used to ensure if we accidentally send more funds we can still withdraw them and they don't get locked

Reply:
I see, your process sounds solid. You're good to go!